feat: enabling others to share their job offers

AceTheCreator opened the pull request:
<!--   Thank you for your contribution. Before you submit the pull request:
1. Follow our contribution guidelines
2. Test your changes and attach their results to the pull request.
3. Update the relevant documentation.
-->

**Description**
This pull request enables others to share their job offers on the AsyncApi website job page
- once the closing date passes, the offer will not be displayed
- instructions on the website on how to publish a job offer on the AsyncApi website.
- Job filtering by category functionality

**Related issue(s)**
<!-- If you refer to a particular issue, provide its number, otherwise, remove this section.
For example, `Resolves #123`, `Fixes #43`, or `See also #33`. The `See also #33` option will not automatically close the issue after the PR merge. -->
Resolves #212

**Quick screen peek**
![Screenshot from 2021-04-07 06-53-29](https://user-images.githubusercontent.com/40604284/113818028-1a19e200-976f-11eb-8915-4a401c0d06b8.png)
![Screenshot from 2021-04-07 06-42-57](https://user-images.githubusercontent.com/40604284/113818148-4a618080-976f-11eb-9329-027fa2f29a33.png)

AceTheCreator committed:
added the job posting functionality

AceTheCreator committed:
adding filter functionality

AceTheCreator committed:
.

AceTheCreator committed:
Merge branch 'master' of https://github.com/AceTheCreator/website-1

AceTheCreator committed:
feat: :lipstick: Enabling others to share their job offers

AceTheCreator committed:
feat: :lipstick: Enabling others to share their job offers

AceTheCreator committed:
feat: :sparkles: Enabling others to share their job offers

feat: Enabling others to share their job offers

AceTheCreator committed:
feat: Enabling others to post jobs on AsyncApi

AceTheCreator committed:
feat: Enabling others to post jobs on AsyncApi website

derberg commented on pages/jobs/ui-ux-dx-designer.md:
```diff
@@ -1,8 +1,8 @@
-category: design
-closingOn: 01/15/2021
+category: Design
+closingOn: 02/15/2021
```
@fmvilas what closing date should we set here?

fmvilas replied:
I don't know yet. Put something far in the future for now.

derberg replied:
@AceTheCreator set june please 😉

AceTheCreator replied:
Ok

AceTheCreator replied:
@derberg and @fmvilas I actually have a friend that fits in this role. Can I refer her?

derberg replied:
@AceTheCreator oh man, definitely! please share it with her and let her apply

derberg commented on pages/jobs/index.js:
```diff
@@ -35,19 +47,52 @@ export default function JobsIndexPage() {
+}
+
+// eslint-disable-next-line no-unused-vars
+export async function getStaticProps(context) {
```
do we need this helper with all those eslint rules disabled? Is it only to read encoded template? why not just having encoded template as a variable directly in code?

AceTheCreator replied:
Hmmm... That's thoughtful of you. I thought it will be too long but yeah I'll do that.

derberg commented on utils/emptyChecker.js:
```diff
@@ -0,0 +1,13 @@
+// Empty object checker
```
do we need to disable those eslint rules?

AceTheCreator replied:
I don't think it really necessary tho. I'll enable them on the next PR.

derberg commented on package.json:
```diff
@@ -3,7 +3,7 @@
   "description": "AsyncAPI website",
   "scripts": {
-    "dev": "node scripts/build-post-list.js && node scripts/build-roadmap.js && next -p 8080",
+    "dev": "node scripts/build-post-list.js && node scripts/build-roadmap.js && next -p 5000",
```
why changing port for dev?

AceTheCreator replied:
Ooops I forgot to revert it. I'm using port 8080, that's why I changed it to 5000. I'll revert it back on my next PR

derberg replied:
what you mean about next PR? why next PR? I noticed you created few PRs already. Why not just push to the branch and get this PR updated?

AceTheCreator replied:
Sorry I meant updated PR 🙂

derberg commented:
@AceTheCreator looks good:

- please adjust PR title to follow conventional commits spec (one of the failing check will give you some hints in logs)
- remove the yarn.lock file from the PR

and look on other comments in the code 🙏🏼

fmvilas commented:
> once the closing date passes, the offer will not be displayed

This website is a static website generated at build time so this is not going to work as you may expect.

AceTheCreator commented:
> This website is a static website generated at build time so this is not going to work as you may expect.

What do you suggest I do @fmvilas?

AceTheCreator committed:
feat: :sparkles: feat: Enabling other to post jobs on AsynApi website

This pull request enables others to share their job offers on the AsyncApi website job page

Resolves: #212

AceTheCreator committed:
Delete yarn.lock

AceTheCreator committed:
feat: :sparkles: feat: Enabling others to share their job offers

This pull request enables others to share their job offers on the AsyncApi website job page

Resolves: #212

AceTheCreator committed:
feat: Enabling others to share their job offers

This pull request enables others to share their job offers on the AsyncApi website job page

Resolves: #212

AceTheCreator committed:
Merge branch 'master' of https://github.com/AceTheCreator/website-1

AceTheCreator committed:
feat: :sparkles: add job posting functionality

AceTheCreator committed:
Merge branch 'master' of https://github.com/AceTheCreator/website-1

AceTheCreator committed:
feat: :zap: add job posting functionality

This pull request enables others to share their job offers on the AsyncApi website job page

Resolves: #212

AceTheCreator commented:
> @AceTheCreator looks good:
> 
> * please adjust PR title to follow conventional commits spec (one of the failing check will give you some hints in logs)
> * remove the yarn.lock file from the PR
> 
> and look on other comments in the code 🙏🏼

@derberg I've updated the PR. kindly review it.

AceTheCreator committed:
changed port back to default

derberg committed:
Merge branch 'master' into master

derberg commented:
why did you add `noreferrer` to all outgoing links? we sometimes look at stats on github to learn how people get into the given repo, from what location. Also I don't see why open collective cannot see that we refer to their website.

any specific reason?

AceTheCreator commented:
> why did you add `noreferrer` to all outgoing links? we sometimes look at stats on github to learn how people get into the given repo, from what location. Also I don't see why open collective cannot see that we refer to their website.
> 
> any specific reason?

I added noreferrer because of security reasons but since you said you sometimes use it for stats on gh, well maybe i should revert it.

derberg commented on pages/jobs/index.js:
```diff
@@ -35,15 +47,30 @@ export default function JobsIndexPage() {
             <p className="mt-3 max-w-2xl mx-auto text-xl leading-7 text-gray-500 sm:mt-4">
               Want to work on a great open-source project with a lovely team and a vibrant community? Come work with us!
```
```suggestion
              Want to work on a great open-source project with a vibrant community? Browse through available job offers!
```

derberg commented on pages/jobs/index.js:
```diff
@@ -35,15 +47,30 @@ export default function JobsIndexPage() {
               Want to work on a great open-source project with a lovely team and a vibrant community? Come work with us!
             </p>
+            <p className="max-w-2xl mx-auto text-md leading-7 text-gray-400">
+            <span role="img">💡</span> Want to post a job? All job posting must be AsyncAPI-related jobs. <a className="ml-1 text-primary-500 hover:text-primary-400" href={`https://github.com/asyncapi/website/new/master/pages/jobs?value=${jobPostUrl}`} target="_blank" rel="noreferrer">Get started now!</a>
```
```suggestion
            <span role="img">💡</span> Want to post a job offer? All job postings must be AsyncAPI-related. <a className="ml-1 text-primary-500 hover:text-primary-400" href={`https://github.com/asyncapi/website/new/master/pages/jobs?value=${jobPostUrl}`} target="_blank" rel="noreferrer">Get started now!</a>
```

derberg commented on pages/jobs/index.js:
```diff
@@ -35,15 +47,30 @@ export default function JobsIndexPage() {
+            <p className="max-w-2xl mx-auto text-md leading-7 text-gray-400">
+            <span role="img">💡</span> Want to post a job? All job posting must be AsyncAPI-related jobs. <a className="ml-1 text-primary-500 hover:text-primary-400" href={`https://github.com/asyncapi/website/new/master/pages/jobs?value=${jobPostUrl}`} target="_blank" rel="noreferrer">Get started now!</a>
+            </p>
+            <p className="max-w-2xl mx-auto text-md leading-7 text-gray-400">Do you want to discuss about your job posting?
```
```suggestion
            <p className="max-w-2xl mx-auto text-md leading-7 text-gray-400">Do you want to discuss your job offer first?
```

derberg reviewed (commented):
some wording improvement.

AceTheCreator commented:
> > why did you add `noreferrer` to all outgoing links? we sometimes look at stats on github to learn how people get into the given repo, from what location. Also I don't see why open collective cannot see that we refer to their website.
> > any specific reason?
> 
I added noreferrer because of security reasons "
The other page may run on the same process as your page. If the other page is running a lot of JavaScript, your page's performance may suffer.
The other page can access your window object with the window.opener property. This may allow the other page to redirect your page to a malicious URL.
" but since you said you sometimes use it for stats on gh, well maybe i should revert it.

derberg commented on components/navigation/Filter.js:
```diff
@@ -0,0 +1,38 @@
+import PropTypes from 'prop-types';
+
+export default function Filter({data, onFilter}) {
+    const [state, setState] = useState('Add filter');
```
when I click on Add filter and select one, the drop-down is not collapsed, it is because the intention is to have multiple selections there?

AceTheCreator replied:
No. I think something went wrong there. Let me fix that and update the PR

AceTheCreator replied:
@derberg I think it has something to do with Tailwind CSS limitations, I'll use a custom CSS

derberg commented on components/navigation/Filter.js:
```diff
@@ -0,0 +1,38 @@
+import PropTypes from 'prop-types';
+
+export default function Filter({data, onFilter}) {
+    const [state, setState] = useState('Add filter');
```
```suggestion
    const [state, setState] = useState('Filter');
```

I think just Filter is fine

AceTheCreator replied:
@derberg I've updated the PR, please kindly review.

AceTheCreator committed:
updated PR

AceTheCreator committed:
Merge branch 'master' of https://github.com/AceTheCreator/website-1

derberg commented on pages/jobs/index.js:
```diff
@@ -33,17 +45,32 @@ export default function JobsIndexPage() {
+              Want to work on a great open-source project with a lovely team and a vibrant community? Browse through available job offers!
+            </p>
+            <p className="max-w-2xl mx-auto text-md leading-7 text-gray-400">
+            <span role="img">💡</span> Want to post a job offer? All job postings must be AsyncAPI-related jobs. <a className="ml-1 text-primary-500 hover:text-primary-400" href={`https://github.com/asyncapi/website/new/master/pages/jobs?value=${jobPostUrl}`} target="_blank">Get started now!</a>
```
sorry, one more suggestion to make it even shorter 😅 

```suggestion
            <span role="img">💡</span> Want to post a job offer? All jobs must be AsyncAPI-related. <a className="ml-1 text-primary-500 hover:text-primary-400" href={`https://github.com/asyncapi/website/new/master/pages/jobs?value=${jobPostUrl}`} target="_blank">Get started now!</a>
```

derberg commented:
> This website is a static website generated at build time so this is not going to work as you may expect.

@fmvilas I think it is fine for now, we redeploy website anyway once a week, so if some job posting that due date passed will be there in the UI for 7days longer, this should not be that bad, or?

Just not sure it make sense for @AceTheCreator to write some additional script for the UI that hides old job postings with some magic on page load. In the end, later it would make more sense to have maybe scheduled github action that checks dates and removes old job posts from repo? So, just a followup issue?

....and please have a look at the preview https://606da5b8e0c59a0007e63de0--asyncapi-website.netlify.app/jobs. I'm supper happy with what we have and we could merge, but, but you know I suck with my sense of aesthetics in UI design 😅

AceTheCreator committed:
Update pages/jobs/index.js

Co-authored-by: Lukasz Gornicki <lpgornicki@gmail.com>

fmvilas commented:
@AceTheCreator @derberg sorry, I should have provided better feedback. It's actually not that bad. You'll see the job posting appearing (server render) and immediately disappearing (client re-render). So there will be a "blink" effect for a second or so. Not that bad.

fmvilas commented on pages/jobs/index.js:
```diff
@@ -33,21 +45,36 @@ export default function JobsIndexPage() {
+              Want to work on a great open-source project with a lovely team and a vibrant community? Browse through available job offers!
+            </p>
+            <p className="max-w-2xl mx-auto text-md leading-7 text-gray-400">
+            <span role="img">💡</span> Want to post a job offer? All jobs must be AsyncAPI-related. <a className="ml-1 text-primary-500 hover:text-primary-400" href={`https://github.com/asyncapi/website/new/master/pages/jobs?value=${jobPostUrl}`} target="_blank">Get started now!</a>
```
What about this?

```suggestion
            <span role="img">💡</span> Want to post a job offer? <a className="ml-1 text-primary-500 hover:text-primary-400" href={`https://github.com/asyncapi/website/new/master/pages/jobs?value=${jobPostUrl}`} target="_blank">Post it now!</a>
```

The disclaimer doesn't need to be here. It can be [here](https://github.com/asyncapi/website/new/master/pages/jobs?value=---%0Atitle%3A%20%27Job%20Title%27%0Adate%3A%20DD%2FMM%2FYYYY%20(current%20date)%0Acategory%3A%20job%20category%0AclosingOn%3A%20MM%2FDD%2FYYYY%0A---%0A%23%23%20About%20the%20team%0ALet%20potential%20employees%20knows%20what%20it%27ll%20feel%20like%20to%20joining%20your%20team%20in%20this%20section%0A%0A%23%23%20TL%3BDR%0AEnter%20list%20of%20perks%20for%20potential%20employees%20in%20this%20section.%0AE.g.%0A*%20%3Amuscle%3A%20Growing%20team%0A*%20%3Ahouse_with_garden%3A%20Fully%20remote%20job%0A*%20%3Amoney_mouth_face%3A%20Great%20salary%20and%20compensation%20package%0A*%20%3Amountain_snow%3A%20Unlimited%20paid%20time%20off%0A%0A%23%23%20About%20the%20job%0ALet%20the%20potential%20employee%20knows%20about%20the%20job%20in%20this%20section%0A%0A%23%23%20About%20youY%0AThe%20Potential%20employees%20would%20love%20to%20know%20more%20about%20you.%20Tell%20them%20more%20about%20you%20in%20this%20section%0A%0A%23%23%20Pay%20and%20benefits%0ALet%20the%20potential%20employees%20knows%20what%20comes%20with%20joining%20your%20team%20in%20this%20section.%0A%0AJoin%20us!%0A).

fmvilas commented on pages/jobs/index.js:
```diff
@@ -33,21 +45,36 @@ export default function JobsIndexPage() {
+            <span role="img">💡</span> Want to post a job offer? All jobs must be AsyncAPI-related. <a className="ml-1 text-primary-500 hover:text-primary-400" href={`https://github.com/asyncapi/website/new/master/pages/jobs?value=${jobPostUrl}`} target="_blank">Get started now!</a>
+            </p>
+            <p className="max-w-2xl mx-auto text-md leading-7 text-gray-400"> Do you want to discuss your job offer first?
+            <a className="ml-1 text-primary-500 hover:text-primary-400" href="https://github.com/asyncapi/website/issues/new" target="_blank">Get started here</a>
```
```suggestion
            <a className="ml-1 text-primary-500 hover:text-primary-400" href="https://github.com/asyncapi/website/issues/new" target="_blank">Get started here.</a>
```

fmvilas commented on templates/job_template.md:
```diff
@@ -0,0 +1,27 @@
+## About the job
+Let the potential employee knows about the job in this section
+
+## About youY
```
```suggestion
## About you
```

fmvilas commented on templates/job_template.md:
```diff
@@ -0,0 +1,27 @@
+category: job category
+closingOn: MM/DD/YYYY
+---
+## About the team
```
```suggestion
## Disclaimer

The job posting must be an AsyncAPI-related job.

## About the team
```

fmvilas commented on css/styles.css:
```diff
@@ -34,4 +34,121 @@ li > p {
\ No newline at end of file
+/*! purgecss end ignore */
+
+#sortbox:checked ~ #sortboxmenu {
```
I know this is going to be painful to read but, can we get rid of this CSS in favor of Tailwind-style classes?

AceTheCreator replied:
It is an old style and it's not relevant anymore. I'll get rid of it now

AceTheCreator replied:
I've removed it

AceTheCreator replied:
I think we're good to go

fmvilas commented on components/navigation/Filter.js:
```diff
@@ -0,0 +1,40 @@
+        onFilter(jobs);
+    }
+    return (
+        <div>
```
I'm adding some margin-bottom here.

```suggestion
        <div class="mb-4">
```

fmvilas reviewed (changes requested):
I'm loving this section so far. Suggested a few implementation details changes.

AceTheCreator committed:
Update pages/jobs/index.js

Co-authored-by: Fran Méndez <fmvilas@gmail.com>

AceTheCreator committed:
Update pages/jobs/index.js

Co-authored-by: Fran Méndez <fmvilas@gmail.com>

AceTheCreator committed:
Update templates/job_template.md

Co-authored-by: Fran Méndez <fmvilas@gmail.com>

AceTheCreator committed:
Update templates/job_template.md

Co-authored-by: Fran Méndez <fmvilas@gmail.com>

AceTheCreator committed:
Update components/navigation/Filter.js

Co-authored-by: Fran Méndez <fmvilas@gmail.com>

AceTheCreator committed:
Update styles.css

fmvilas commented on components/navigation/Filter.js:
```diff
@@ -0,0 +1,40 @@
+    }
+    return (
+        <div class="mb-4">
+            <div className="select-box inline-flex w-56 max-w-sm justify-between">
```
I just recalled we already have a Select component: https://github.com/asyncapi/website/blob/master/components/form/Select.js. Please use this one. If you're curious how it looks like, it's the one here: https://www.asyncapi.com/generator.

AceTheCreator replied:
Ohh... So I should get rid of the one here right? I'll replace it and update the PR

AceTheCreator replied:
But mine looks more fancy Lol 😂. I'll replace it then.

AceTheCreator replied:
Thanks for the review tho

fmvilas replied:
haha! It looks better indeed. Feel free to improve the existing one 😄 There's something that's also better in the existing one and it's that it uses a native `select` HTML tag instead of a div.

fmvilas commented on components/navigation/Filter.js:
```diff
@@ -0,0 +1,40 @@
+    const catgeories = [];
+    data.map((data) => {
+        catgeories.push(data.category)
+    });
```
```suggestion
    const catgeories = data.map(data => data.category);
```

fmvilas commented on css/styles.css:
```diff
@@ -34,4 +34,117 @@ li > p {
+    -webkit-transform: scaleY(0);
+            transform: scaleY(0);
+  }
+}
```
These styles are not needed anymore.

AceTheCreator replied:
Ok

fmvilas commented on pages/jobs/index.js:
```diff
@@ -33,21 +45,36 @@ export default function JobsIndexPage() {
-            }
-            </ul>
+          <div className="text-center">
+            {emptyChecker(posts) ? <div className="flex content-center justify-center">
```
What about this?

```suggestion
            {!Object.keys(posts).length ? <div className="flex content-center justify-center">
```

AceTheCreator replied:
Wow!!! That's cool, then i don't need the empty checker helper.

AceTheCreator replied:
But do you think it's better than the empty checker helper function?

fmvilas replied:
At least it's self-contained. I mean, it doesn't need to import something on another file. If you want to make it easier to understand you can probably define a variable before the return like this:

```js
const hasPosts = Object.keys(posts).length;
```

Then you can use it like `{!hasPosts ? <div...`

AceTheCreator replied:
Ok cool

AceTheCreator replied:
@fmvilas I've updated the PR, please kindly review

fmvilas reviewed (changes requested):
Left some comments on how to simplify the code.

AceTheCreator committed:
Update components/navigation/Filter.js

Co-authored-by: Fran Méndez <fmvilas@gmail.com>

AceTheCreator committed:
used the existing select component and got rid of the custom one

fmvilas commented on components/navigation/Filter.js:
```diff
@@ -0,0 +1,33 @@
+            text : data.category
+        }
+    });
+    console.log(categories);
```
```suggestion

```

AceTheCreator replied:
Lol. I forgot to remove them before pushing 😀

fmvilas commented on components/navigation/Filter.js:
```diff
@@ -0,0 +1,33 @@
+
+export default function Filter({data, onFilter}) {
+    const [state, setState] = useState('Filter');
+    console.log(state);
```
```suggestion

```

fmvilas reviewed (changes requested):
Last things, I promise :)

AceTheCreator committed:
Update components/navigation/Filter.js

Co-authored-by: Fran Méndez <fmvilas@gmail.com>

AceTheCreator committed:
Update components/navigation/Filter.js

Co-authored-by: Fran Méndez <fmvilas@gmail.com>

fmvilas reviewed (approved):
Great stuff!! 🙌

AceTheCreator commented:
> Great stuff!!

Thanks. Looking forward to more contributions.
